src/main/java/com/google/devtools/build/lib/rules/cpp/CppHelper.java

```diff
@@ -903,7 +918,23 @@ private static Artifact createEmptyDefFileAction(RuleContext ruleContext) {
             ruleContext.getLabel().getName()
                 + ".gen.empty"
                 + Iterables.getOnlyElement(CppFileTypes.WINDOWS_DEF_FILE.getExtensions()));
-    ruleContext.registerAction(FileWriteAction.create(ruleContext, trivialDefFile, "", false));
+    String fileContent = "";
+    CppConfiguration cppConfiguration = ruleContext.getFragment(CppConfiguration.class);
+    if(cppConfiguration.isRenameDLL()) {
+      String dllName = ruleContext.getLabel().getName();
+      String dllPath = ruleContext.getRepository() 
+          + "/" 
+          + ruleContext.getPackageDirectory();
+      String dllHash = Hashing.sha256()
+          .hashString(dllPath, Charsets.UTF_8)
+          .toString();
+      String dllNameWithoutExtension = dllName
+          .replaceAll(".dll", "");
```

Comment:
We have to think about the situation where the extension is hardcorded (i.e name = "utils.so"). Maybe hash_name.dll is better than name_hash.dll

Reply:
Yes. `createEmptyDefFileAction` is used by both CcBinary and CcLibrary. We don't actually want to change anything for CcBinary. You'll have to distinguish them. Maybe add an optional argument to specify the DLL name?

And to generate the hash, can you do something similar to:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/test/TestStrategy.java;l=288?q=getTmpDirName&ss=bazel%2Fbazel

Using the Fingerprint class seems to be a nicer approach. And to avoid [the long path issues](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#maximum-path-length-limitation) on Windows, can you truncate the hash to 16 or maybe 8 characters?